server/main_server/src/agent_main.cpp

```diff
@@ -0,0 +1,1326 @@
+                log_agent::error(PASS(ret).result());
+                sendVersion(net_obj, SERVER_NEGOTIATION_ERROR, 0, nullptr, 0);
+                cleanupAndExit(ret.code());
+            }
+            else {
+                // copy negotiation results to comm for action by network objects
+                std::snprintf(rsComm.negotiation_results, sizeof(rsComm.negotiation_results), "%s", neg_results.c_str());
+            }
+        }
+
+        // send the server version and status as part of the protocol. Put
+        // rsComm.reconnPort as the status
+        ret = sendVersion(net_obj, status, rsComm.reconnPort, rsComm.reconnAddr, rsComm.cookie);
+
+        if (!ret.ok()) {
+            log_agent::error(PASS(ret).result());
+            sendVersion(net_obj, SYS_AGENT_INIT_ERR, 0, nullptr, 0);
+            cleanupAndExit(status);
+        }
+
+        create_agent_pid_file_for_ips(rsComm, _created_at);
+
+        // call initialization for network plugin as negotiated
+        irods::network_object_ptr new_net_obj;
+        ret = irods::network_factory(&rsComm, new_net_obj);
+        if (!ret.ok()) {
+            log_agent::error(PASS(ret).result());
+            return 1;
+        }
+
+        ret = sockAgentStart(new_net_obj);
+        if (!ret.ok()) {
+            log_agent::error(PASS(ret).result());
+            return 1;
+        }
+
+        const auto cleanup_and_free_rsComm_members = [&rsComm] {
+            cleanup();
+            if (rsComm.thread_ctx) {
+                std::free(rsComm.thread_ctx); // NOLINT(cppcoreguidelines-no-malloc, cppcoreguidelines-owning-memory)
+            }
+            if (rsComm.auth_scheme) {
+                std::free(rsComm.auth_scheme); // NOLINT(cppcoreguidelines-no-malloc, cppcoreguidelines-owning-memory)
+            }
+        };
+
+        new_net_obj->to_server(&rsComm);
+        status = agentMain(rsComm);
+
+        // call initialization for network plugin as negotiated
+        ret = sockAgentStop( new_net_obj );
+        if (!ret.ok()) {
+            log_agent::error(PASS(ret).result());
+            cleanup_and_free_rsComm_members();
+            return 1;
+        }
+
+        new_net_obj->to_server(&rsComm);
+
+        cleanup_and_free_rsComm_members();
+
+        // clang-format off
+        (0 == status)
+            ? log_agent::debug("Agent [{}] exiting with status = {}", getpid(), status)
+            : log_agent::error("Agent [{}] exiting with status = {}", getpid(), status);
+        // clang-format on
+
+#if __has_feature(address_sanitizer) || defined(__SANITIZE_ADDRESS__)
+        // This function must be called here due to the use of _exit() (just below). Address Sanitizer (ASan)
+        // relies on std::atexit handlers to report its findings. _exit() does not trigger any of the handlers
+        // registered by ASan, therefore, we manually run ASan just before the agent exits.
+        __lsan_do_leak_check();
+#endif
```

Comment:
This may be unnecessary now that the implementation is more sane.

Reply:
I'm pretty sure this isn't necessary since we no longer use `_exit()` in the agents.

Will verify first though.

Reply:
This isn't needed anymore. I've confirmed ASan detects memory leaks in the agents without this block.

Will remove.

Reply:
Done.